python-frontend/src/main/java/org/sonar/python/types/v2/TypeCheckBuilder.java

```diff
@@ -125,4 +138,94 @@ public TriBool test(PythonType pythonType) {
       return pythonType.equals(expectedType) ? TriBool.TRUE : TriBool.FALSE;
     }
   }
+
+  record IsInstanceOfPredicate(PythonType expectedType)  implements TypePredicate {
+
+    @Override
+    public TriBool test(PythonType pythonType) {
+      if (expectedType instanceof ClassType expectedClassType) {
+        if (pythonType instanceof ObjectType objectType) {
+          if (objectType.type() instanceof ClassType classType) {
+            // when the checking type is an ObjectType of a ClassType
+            return isClassInheritedFrom(classType, expectedClassType);
+          } else if (objectType.type() instanceof UnionType unionType) {
+            // when the checking type is an ObjectType of a UnionType
+            return isObjectOfUnionTypeInstanceOf(expectedClassType, unionType);
+          }
+        } else if (pythonType instanceof UnionType unionType) {
+          // when the checking type is a UnionType
+          return isUnionTypeInstanceOf(unionType);
+        }
+      }
+      return TriBool.UNKNOWN;
+    }
+
+    private static TriBool isObjectOfUnionTypeInstanceOf(ClassType expectedClassType, UnionType unionType) {
+      var results = unionType.candidates()
+        .stream()
+        .map(classType -> isClassInheritedFrom(classType, expectedClassType))
+        .distinct()
+        .toList();
+
+      if (results.size() > 1) {
+        return TriBool.UNKNOWN;
+      } else {
+        return results.get(0);
+      }
+    }
+
+    private TriBool isUnionTypeInstanceOf(UnionType unionType) {
+      var candidatesResults = unionType.candidates()
+        .stream()
+        .map(this::test)
+        .distinct()
+        .toList();
+
+      if (candidatesResults.size() != 1) {
+        return TriBool.UNKNOWN;
+      } else {
+        return candidatesResults.get(0);
+      }
+    }
+
+    private static TriBool isClassInheritedFrom(PythonType classType, ClassType expectedClassType) {
+      if (classType == expectedClassType) {
+        return TriBool.TRUE;
+      }
+
+      var types = collectTypes(classType);
+
+      if (types.contains(expectedClassType)) {
+        return TriBool.TRUE;
+      } else if (types.contains(PythonType.UNKNOWN)) {
+        return TriBool.UNKNOWN;
+      } else {
+        return TriBool.FALSE;
+      }
+    }
+
+    private static Set<PythonType> collectTypes(PythonType type) {
+      var result = new HashSet<PythonType>();
+      collectTypes(type, result);
+      return result;
+    }
+
+    private static void collectTypes(PythonType type, Set<PythonType> result) {
+      result.add(type);
+      if (type instanceof ClassType classType) {
```

Comment:
Isn't it possible to have a `UnionType` as parent somewhere? Then shouldn't we consider them?
Also, if anything unexpected is in the type hierarchy (i.e not a `ClassType`), how do we make sure we don't miss it?


```
class A: ...
if x:
    class B(A):
        def foo(self): ...
else:
    class B(A):
        def bar(self): ...

class C(B): ...  # checking B instance of A?
```

I think the only reason why it would currently work is because we may not correctly create a `UnionType` for `B`. But if we do, this code would probably miss it.